File: src/backend/codecc/core/defect/biz-defect/src/main/java/com/tencent/bk/codecc/defect/dao/mongotemplate/BuildDefectDao.java
package com.tencent.bk.codecc.defect.dao.mongotemplate;

import com.tencent.bk.codecc.defect.model.BuildDefectEntity;
import com.tencent.bk.codecc.defect.model.CCNDefectEntity;
import com.tencent.bk.codecc.defect.model.CheckerConfigEntity;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections.CollectionUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.BulkOperations;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.stereotype.Repository;

import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * 描述
 *
 * @version V2.0
 * @date 2020/04/16
 */
@Slf4j
@Repository
public class BuildDefectDao
{
    @Autowired
    private MongoTemplate mongoTemplate;

    public void upsertByFilePath(Collection<BuildDefectEntity> buildDefectEntityList)
    {
        BulkOperations ops = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, BuildDefectEntity.class);
        if (CollectionUtils.isNotEmpty(buildDefectEntityList))
        {
            for (BuildDefectEntity buildDefectEntity : buildDefectEntityList)
            {
                Query query = new Query();
                query.addCriteria(Criteria.where("task_id").is(buildDefectEntity.getTaskId())
                        .and("tool_name").is(buildDefectEntity.getToolName())
                        .and("build_id").is(buildDefectEntity.getBuildId())
                        .and("file_path").is(buildDefectEntity.getFilePath()));
                Update update = new Update();
                update.set("task_id", buildDefectEntity.getTaskId())
                        .set("tool_name", buildDefectEntity.getToolName())
                        .set("build_id", buildDefectEntity.getBuildId())
                        .set("build_num", buildDefectEntity.getBuildNum())
                        .set("file_path", buildDefectEntity.getFilePath())
                        .set("file_defect_ids", buildDefectEntity.getFileDefectIds());
                ops.upsert(query, update);
            }
            ops.execute();
        }
    }


    public void upsertByDefectId(Collection<BuildDefectEntity> buildDefectEntityList)
    {
        BulkOperations ops = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, BuildDefectEntity.class);
        if (CollectionUtils.isNotEmpty(buildDefectEntityList))
        {
            for (BuildDefectEntity buildDefectEntity : buildDefectEntityList)
            {
                Query query = new Query();
                query.addCriteria(Criteria.where("task_id").is(buildDefectEntity.getTaskId())
                        .and("tool_name").is(buildDefectEntity.getToolName())
                        .and("build_id").is(buildDefectEntity.getBuildId())
                        .and("defect_id").is(buildDefectEntity.getDefectId()));
                Update update = new Update();
                update.set("task_id", buildDefectEntity.getTaskId())
                        .set("tool_name", buildDefectEntity.getToolName())
                        .set("build_id", buildDefectEntity.getBuildId())
                        .set("build_num", buildDefectEntity.getBuildNum())
                        .set("defect_id", buildDefectEntity.getDefectId());
                ops.upsert(query, update);
            }
            ops.execute();
        }
    }
}

File: src/backend/codecc/core/apiquery/biz-apiquery/src/main/kotlin/com/tencent/bk/codecc/apiquery/defect/dao/BuildDefectDao.kt
package com.tencent.bk.codecc.apiquery.defect.dao

import com.tencent.bk.codecc.apiquery.defect.model.BuildDefectIdsModel
import com.tencent.bk.codecc.apiquery.defect.model.BuildDefectModel
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Sort
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.aggregation.Aggregation
import org.springframework.data.mongodb.core.aggregation.AggregationOptions
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import org.springframework.stereotype.Repository

@Repository
class BuildDefectDao @Autowired constructor(
    private val defectMongoTemplate: MongoTemplate
) {

    fun findByTaskIdToolNameAndBuildId(
        taskIdList: List<Long>,
        toolName: String,
        buildId: String
    ): List<BuildDefectModel> {
        val query = Query.query(Criteria.where("task_id").`in`(taskIdList)
            .and("tool_name").`is`(toolName)
            .and("build_id").`is`(buildId))
        return defectMongoTemplate.find(query, BuildDefectModel::class.java, "t_build_defect")
    }

    fun findByTaskIdToolNameAndBuildIdOrderByFilePath(
        taskIdList: List<Long>,
        toolName: String,
        buildId: String,
        pageNum: Int?,
        pageSize: Int?,
        sortField : String?,
        sortType : String?
    ): List<BuildDefectIdsModel> {
        val match = Aggregation.match(
            Criteria.where("task_id").`in`(taskIdList)
                .and("tool_name").`is`(toolName)
                .and("build_id").`is`(buildId)
        )

        val unwind = Aggregation.unwind("file_defect_ids")
        val project = Aggregation.project("file_defect_ids")

        val skip = Aggregation.skip((((pageNum ?: 1) - 1) * (pageSize ?: 100)).toLong())
        val limit = Aggregation.limit((pageSize ?: 100).toLong())
        //允许磁盘操作
        val options = AggregationOptions.Builder().allowDiskUse(true).build()
        val agg = if(!sortField.isNullOrBlank() && !sortType.isNullOrBlank()){
            val sort = Aggregation.sort(Sort.by(Sort.Direction.fromString(sortType), sortField))
            Aggregation.newAggregation(match, sort, unwind, project, skip, limit).withOptions(options)
        } else {
            Aggregation.newAggregation(match, unwind, project, skip, limit).withOptions(options)
        }



        return defectMongoTemplate.aggregate(agg, "t_build_defect", BuildDefectIdsModel::class.java).mappedResults
    }
}
